Components: refactor `Navigator*` folder structure, rename `Navigator` to `NavigatorProvider`

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

_This is part of a series of smaller PRs that aim at refining the family of `Navigator` components and is part of a larger effort tracked in #34907._

This PR slightly refactors the folder structure of the `Navigator*` components family, following the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md):

- `Navigator` has been renamed to `NavigatorProvider` (also to avoid confusion with the `useNavigator` hook, which could be otherwise mistaken for [the hook of the `Navigator` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#components--hooks))
- Code for `NavigatorProvider` and `NavigatorScreen` have been moved to subfolders, respectively `navigator-provider` and `navigator-screen`
- Added a separate README for `NavigatorScreen`
- Updated all references to `Navigator` to now reference `NavigatorProvider`


Future follow-up PRs will introduce unit tests and refactor the family of components to TypeScript.


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

There are no runtime changes.

- The project builds
- Storybook example works as expected
- Global Styles sidebar works as expected

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ciampo committed:
Rename `Navigator` component to `NavigatorRoot`, move code under `navigator-root` folder

ciampo committed:
Move `NavigatorScreen` under `navigator-screen` folder, add README

youknowriad commented on packages/components/src/index.js:
```diff
@@ -97,7 +97,7 @@ export { default as __experimentalNavigationGroup } from './navigation/group';
 export { default as __experimentalNavigationMenu } from './navigation/menu';
 export {
-	Navigator as __experimentalNavigator,
+	NavigatorRoot as __experimentalNavigatorRoot,
```
To be honest I don't like the name `NavigatorRoot`. I feel it can be confused with the "Root Screen". In other places in the UI Components we use `Provider` so maybe `NavigatorProvider` would be better for me. That said, don't take my word for it, I know I'm not great at naming things :P

ciampo replied:
Naming things is hard, and doing so in a project like Gutenberg is even harder! `NavigatorProvider` sounds good to me, I'll go ahead and make changes

ciampo committed:
Rename `NavigatorRoot` to  `NavigatorProvider`

ciampo commented:
- Addressed feedback by renaming `NavigatorRoot` to `NavigatorProvider`
- Rebased on top of `trunk` to include latest changes from #35163

@youknowriad , this PR is ready for another round of review

youknowriad reviewed (approved):
LGTM 👍

ciampo commented:
~~Hey @ntsekouras , I'm afraid this PR was not ready to be merged — it should have been rebased after #35109 got merged in order to rename `Navigator` to `NavigatorProvider` also in there. I'm opening a PR to fix that~~

Edit: actually nevermind, my bad. Looks like the 2 parallel PRs didn't cause any conflicts. Sorry for the ping!
